warn-stranded.lua

```diff
@@ -0,0 +1,581 @@
+--   deserializedIgnores is optional but allows us to only call deserialize once like an explicit cache.
+local function unitIgnored(unit, deserializedIgnores)
+    local ignores = deserializedIgnores or deserializeIgnoredUnits()
+
+    for index, id in ipairs(ignores) do
+        if tonumber(id) == unit.id then
+            return true, index
+        end
+    end
+
+    return false
+end
+
+-- Check for and potentially add [IGNORED] to text.
+--   Optional deserializedIgnores allows us to call deserialize once for a group of operations
+local function addIgnored(text, unit, deserializedIgnores)
+    if unitIgnored(unit, deserializedIgnores) then
+        return text..'[IGNORED] '
+    end
+
+    return text
+end
+
+-- Uses persistent API. Toggles a unit's ignored status by deserializing 'warnStrandedIgnore' key
+--   then serializing the resulting table after the toggle.
+-- Optional cache parameter could affect data integrity. Make sure you don't need data reloaded
+--   before using it. Calling several times in a row can use the return result of the function
+--   as input to the next call.
+local function toggleUnitIgnore(unit, deserializedIgnores)
+    local ignores = deserializedIgnores or deserializeIgnoredUnits()
+    local is_ignored, index = unitIgnored(unit, ignores)
+
+    if is_ignored then
+        table.remove(ignores, index)
+    else
+        table.insert(ignores, unit.id)
+    end
+
+    dfhack.persistent.delete('warnStrandedIgnore')
+    dfhack.persistent.save({key = 'warnStrandedIgnore', value = table.concat(ignores, ' ')})
+
+    return ignores
+end
+
+-- Does the usual GUI pattern when groups can be in a partial state
+--   Will ignore everything, unless all units in group are already ignored
+--   If all units in the group are ignored, then it will unignore all of them
+local function toggleGroup(groups, groupNumber)
+    local ignored = deserializeIgnoredUnits()
+
+    if groupNumber > #groups then
+        print('Group '..groupNumber..' does not exist')
+        return false
+    end
+
+    if groups[groupNumber]['mainGroup'] then
+        print('Group '..groupNumber..' is the main group of dwarves. Cannot toggle.')
+        return false
+    end
+
+    local group = groups[groupNumber]
+
+    local allIgnored = true
+    for _, unit in ipairs(group['units']) do
+        if not unitIgnored(unit, ignored) then
+            allIgnored = false
+            goto process
+        end
+    end
+    ::process::
+
+    for _, unit in ipairs(group['units']) do
+        local isIgnored = unitIgnored(unit, ignored)
+
+        if allIgnored == isIgnored then
+            ignored = toggleUnitIgnore(unit, ignored)
+        end
+    end
+
+    return true
+end
+
+-- ===============================================================
+--                   Graphical Interface
+-- ===============================================================
+warning = defclass(warning, gui.ZScreenModal)
```

Comment:
Classes should start with an upper case letter (I'm aware that `warn-starving` fails to follow that pattern; I'll have to update the code there)